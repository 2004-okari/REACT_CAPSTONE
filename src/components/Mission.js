import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { toggleReservation } from '../redux/mission/missionSlice';
import './mission.css';

function Mission() {
  const dispatch = useDispatch();
  const missions = useSelector((state) => state.missions.missions);
  const { status, error } = useSelector((state) => state.missions);

  const handleReservationToggle = (missionId) => {
    dispatch(toggleReservation(missionId));
  };

  return (
    <>
      {status === 'loading' ? (<span className="loader">Load&nbsp;ng</span>) : ''}
      {status === 'failed' ? (
        <span>
          Error:
          {error}
        </span>
      ) : ''}
      <table className="table">
        <thead className="thead">
          <tr className="thead-row">
            <td className="thead-title">Mission</td>
            <td className="thead-title">Description</td>
            <td className="thead-title">Status</td>
            <td className="thead-title" aria-label="join" />
          </tr>
        </thead>
        <tbody className="tbody">
          {missions.map((mission) => (
            <tr key={mission.mission_id} className="tbody-row">
              <td className="mission-text">{mission.mission_name}</td>
              <td className="mission-text">{mission.description}</td>
              <td className="btn-status">{!mission.reserved ? (<p className="not">NOT A MEMBER</p>) : (<p className="member">Active member</p>)}</td>
              <td className="mission-text"><button className={mission.reserved ? 'cancel-btn' : 'reserve-btn'} type="button" onClick={() => handleReservationToggle(mission.mission_id)}>{!mission.reserved ? 'Join a mission' : 'Leave mission'}</button></td>
            </tr>
          ))}
        </tbody>
      </table>
    </>
  );
}

export default Mission;
