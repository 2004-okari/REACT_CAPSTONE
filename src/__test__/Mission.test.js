import React from 'react';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import Mission from '../components/Mission';
import '@testing-library/jest-dom/extend-expect';

const mockStore = configureStore([]);

test('renders mission component', () => {
  const initialState = {
    missions: {
      missions: [
        {
          mission_id: 'mission1',
          mission_name: 'Mission 1',
          description: 'Mission 1 description',
          reserved: false,
        },
      ],
      status: 'idle',
      error: null,
    },
  };
  const store = mockStore(initialState);
  render(
    <Provider store={store}>
      <Mission />
    </Provider>,
  );

  const missionNameElement = screen.getByText('Mission 1');
  const missionDescriptionElement = screen.getByText('Mission 1 description');
  expect(missionNameElement).toBeInTheDocument();
  expect(missionDescriptionElement).toBeInTheDocument();
});

test('mission component matches snapshot', () => {
  const initialState = {
    missions: {
      missions: [
        {
          mission_id: 'mission1',
          mission_name: 'Mission 1',
          description: 'Mission 1 description',
          reserved: false,
        },
      ],
      status: 'idle',
      error: null,
    },
  };
  const store = mockStore(initialState);

  const tree = renderer.create(
    <Provider store={store}>
      <Mission />
    </Provider>,
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
