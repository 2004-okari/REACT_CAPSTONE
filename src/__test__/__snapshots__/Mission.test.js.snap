// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mission component matches snapshot 1`] = `
<table
  className="table"
>
  <thead
    className="thead"
  >
    <tr
      className="thead-row"
    >
      <td
        className="thead-title"
      >
        Mission
      </td>
      <td
        className="thead-title"
      >
        Description
      </td>
      <td
        className="thead-title"
      >
        Status
      </td>
      <td
        aria-label="join"
        className="thead-title"
      />
    </tr>
  </thead>
  <tbody
    className="tbody"
  >
    <tr
      className="tbody-row"
    >
      <td
        className="mission-text"
      >
        Mission 1
      </td>
      <td
        className="mission-text"
      >
        Mission 1 description
      </td>
      <td
        className="btn-status"
      >
        <p
          className="not"
        >
          NOT A MEMBER
        </p>
      </td>
      <td
        className="mission-text"
      >
        <button
          className="reserve-btn"
          onClick={[Function]}
          type="button"
        >
          Join a mission
        </button>
      </td>
    </tr>
  </tbody>
</table>
`;
